core/transaction_pool/impl/transaction_pool_impl.cpp

```diff
@@ -24,200 +24,278 @@ namespace kagome::transaction_pool {
     BOOST_ASSERT_MSG(moderator_ != nullptr, "moderator is nullptr");
   }
 
-  outcome::result<void> TransactionPoolImpl::submitOne(Transaction t) {
-    return submit({t});
+  outcome::result<void> TransactionPoolImpl::submitOne(Transaction &&tx) {
+    return submitOne(
+        std::make_shared<Transaction>(std::forward<Transaction>(tx)));
   }
 
   outcome::result<void> TransactionPoolImpl::submit(
-      std::vector<Transaction> ts) {
-    for (auto &t : ts) {
-      auto hash = t.hash.toHex();
-      if (imported_hashes_.find(hash) != imported_hashes_.end()) {
-        return TransactionPoolError::ALREADY_IMPORTED;
-      }
-      for (auto &tag : t.provides) {
-        provided_tags_by_.insert({tag, t.hash});
-      }
-      imported_hashes_.insert(hash);
-      waiting_queue_.push_back(WaitingTransaction{std::move(t)});
+      std::vector<Transaction> txs) {
+    for (auto &tx : txs) {
+      OUTCOME_TRY(submitOne(std::make_shared<Transaction>(std::move(tx))));
+    }
+
+    return outcome::success();
+  }
+
+  outcome::result<void> TransactionPoolImpl::submitOne(
+      const std::shared_ptr<Transaction> &tx) {
+    if (auto [_, ok] = _importedTxs.emplace(tx->hash, tx); !ok) {
+      return TransactionPoolError::TX_ALREADY_IMPORTED;
+    }
+
+    auto processResult = processTx(tx);
+    if (processResult.has_error()
+        && processResult.error() == TransactionPoolError::POOL_IS_FULL) {
+      _importedTxs.erase(tx->hash);
+    }
+
+    return processResult;
+  }
+
+  outcome::result<void> TransactionPoolImpl::processTx(
+      const std::shared_ptr<Transaction> &tx) {
+    if (checkForReady(tx)) {
+      OUTCOME_TRY(processTxAsReady(tx));
+    } else {
+      OUTCOME_TRY(processTxAsWaiting(tx));
     }
+    return outcome::success();
+  }
 
-    updateReady();
+  outcome::result<void> TransactionPoolImpl::processTxAsReady(
+      const std::shared_ptr<Transaction> &tx) {
+    if (hasSpaceInReady()) {
+      setReady(tx);
+    } else {
+      OUTCOME_TRY(ensureSpaceInWaiting());
 
-    enforceLimits();
+      postponeTx(tx);
+    }
 
     return outcome::success();
   }
 
+  bool TransactionPoolImpl::hasSpaceInReady() const {
+    return _readyTxs.size() < limits_.max_ready_num;
+  }
+
+  void TransactionPoolImpl::postponeTx(const std::shared_ptr<Transaction> &tx) {
+    _postponedTxs.push_back(tx);
+  }
+
+  outcome::result<void> TransactionPoolImpl::processTxAsWaiting(
+      const std::shared_ptr<Transaction> &tx) {
+    OUTCOME_TRY(ensureSpaceInWaiting());
+
+    addTxAsWaiting(tx);
+
+    return outcome::success();
+  }
+
+  outcome::result<void> TransactionPoolImpl::ensureSpaceInWaiting() const {
+    if (_importedTxs.size() - _readyTxs.size() > limits_.max_waiting_num) {
+      return TransactionPoolError::POOL_IS_FULL;
```

Comment:
Not certainly in that way. Used two different limit for ready and wainitg transaction. I suggest to use `capacity` instead `waiting_limit`. It is as sum of ready and waiting linits.
It will be changed by additional commit.